### PR TITLE
JetBrains: Add more comprehensive autocomplete telemetry

### PR DESCRIPTION
**Pull request by vdavid**
Closes https://github.com/sourcegraph/cody/issues/123

Changes:
- Added `latency` and `displayDuration` arguments for `CodyJetBrainsPlugin:completion:suggested` event logging
- Added comments to various places as I discovered the codebase.

The other two completion-related events (`completion:started` and `completion:accepted`) were already in place at the right locations. `completion:suggested` was triggered too early, I had to move it to a much later point where the timing information was available.

Caveats:
- Did not add any other arguments for now. I need to check the dashboard to make sure this is enough, but honestly, I didn't want to do a lot of changes before the Agent-related PRs are merged

Note for the reviewer: if the docs changes are too much, you can review this commit by commit. But it's probably not worth it, so I've left comments on the key points; the rest are just added comments.

## Test plan

Made sure that:
- Autocompletions still work
- We're tracking the latency and display information
- That logging has roughly the same shape as the logs in VS Code (see the [issue](https://github.com/sourcegraph/cody/issues/123) comments for a detailed list)

[3-min Loom](https://www.loom.com/share/703b0c41bbfd4a3c94ea2acd6f432fa1?sid=512b70c4-dd49-46ce-8090-5a6ed6f033f2) where I demo this

### CODE REVIEW

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/AutoCompleteTextAtCaret.java [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import com.intellij.openapi.editor.Caret;
 import org.jetbrains.annotations.NotNull;
 
-class AutoCompleteTextAtCaret {
+public class AutoCompleteTextAtCaret {
```
Comment (vdavid) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/AutoCompleteTextAtCaret.java:6]: To fix a warning about the visibility of this class

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java [hunk 2]
```diff
@@ -35,14 +35,34 @@ public class CodyAutoCompleteManager {
...
+        .ifPresent(
+            p -> {
+              if (currentAutocompletion != null
+                  && currentAutocompletion.getStatus()
+                      != AutocompletionStatus.TRIGGERED_NOT_DISPLAYED) {
+                currentAutocompletion.markCompletionHidden();
+                GraphQlLogger.logAutocompleteSuggestedEvent(
+                    p,
+                    currentAutocompletion.getLatencyMs(),
+                    currentAutocompletion.getDisplayDurationMs());
```
Comment (vdavid) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java:57]: This is where we do the new logging!

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java [hunk 3]
```diff
@@ -136,12 +187,11 @@ private CompletableFuture<Void> triggerAutoCompleteAsync(
                           /* Clear existing completions */
                           this.clearAutoCompleteSuggestions(editor);
 
-                          /* Log the event */
-                          Optional.ofNullable(editor.getProject())
-                              .ifPresent(
-                                  p -> GraphQlLogger.logCodyEvent(p, "completion", "suggested"));
```
Comment (vdavid) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java:142]: We used to do the logging here, but it was too early for the current purpose; we have no timing info at this point yet. Instead, we just mark that this was the time when we displayed the completion.

File: client/jetbrains/src/main/java/com/sourcegraph/cody/recipes/SimpleRecipeAction.java [hunk 4]
```diff
@@ -33,7 +33,7 @@ public void actionPerformed(@NotNull AnActionEvent e) {
 
   public void executeRecipeWithPromptProvider(
       @NotNull UpdatableChat updatableChat, @NotNull Project project) {
-    GraphQlLogger.logCodyEvents(project, this.getActionComponentName(), "clicked");
+    GraphQlLogger.logCodyEvent(project, this.getActionComponentName(), "clicked");
```
Comment (vdavid) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/recipes/SimpleRecipeAction.java:36]: Small cleanup

File: client/jetbrains/src/main/java/com/sourcegraph/telemetry/GraphQlLogger.java [hunk 5]
```diff
@@ -16,53 +16,55 @@
 public class GraphQlLogger {
   private static final Logger logger = Logger.getInstance(GraphQlLogger.class);
 
-  public static void logInstallEvent(Project project, Consumer<Boolean> callback) {
```
Comment (vdavid) [inline client/jetbrains/src/main/java/com/sourcegraph/telemetry/GraphQlLogger.java:19]: Refactored the internals of this class to accommodate an elegant `logAutocompleteSuggestedEvent`.

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/Autocompletion.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package com.sourcegraph.cody.autocomplete;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A class that stores the state and timing information of an autocompletion.
+ */
+public class Autocompletion {
+  private long completionTriggeredTimestampMs;
```
Comment (olafurpg) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/Autocompletion.java:9]: FYI: we can use `java.time.Instant.now()` to capture time stamps and `java.time.Duration.between(Instant,Instant)` to measure duration between timestamps. Using `Instant` avoids the need for `TimestampMs` suffix.

Not a blocking comment.

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/Autocompletion.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
+package com.sourcegraph.cody.autocomplete;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A class that stores the state and timing information of an autocompletion.
+ */
+public class Autocompletion {
```
Comment (olafurpg) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/Autocompletion.java:8]: nit: should this be named `AutocompleteEvent` or `AutocompleteTelemetry` or `AutocompleteVisibility`?

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java [hunk 8]
```diff
@@ -96,6 +130,23 @@ public void triggerAutoComplete(@NotNull Editor editor, int offset) {
     }
   }
 
+  /**
+   * Asynchronously triggers auto-complete for the given editor and offset. Details: - Calls
```
Comment (olafurpg) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java:134]: In the future, I think it's best to try and open separate PRs for refactorings/documentation and functional changes.

I'm fine merging this PR as is to unblock the telemetry improvements but I'm generally skeptical of having such large docstrings that repeat the implementation in natural languages because it quickly gets outdated (causing more harm than help). You can always use Cody to generate a natural language summary on the fly.

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java (hunk 3 above)
Comment (olafurpg) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java:142]: Have you confirmed with product what events they are tracking for IntelliJ? Removing events can influence the dashboard, making it impossible to see trends before/after this change.

**Review by olafurpg — APPROVED** [review]
LGTM 👍🏻 The only semi-blocking comment is about the removal of the event because it's a one-way decision that can have unintended negative consequences for metrics tracking.

Left several non-blocking comments because it's urgent we ship the new telemetry live ASAP.

In the future, please try to keep PRs more focused on the core functional change and open separate PRs for code cleanups. Docstrings are most effective IMO when they document something that is not obvious from reviewing the code itself, for example *why* we do something unusual or adding a docstring to package-info.java to explain what kind of code should live in that package (or what kind of code should not live in that package).

**Comment by olafurpg** [comment]
You can fix the failing formatting CI job with `./gradlew spotlessApply`

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java (hunk 3 above)
Comment (vdavid) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java:142]: I didn't remove this; I just moved the same logging to a later point in time—a behavior that matches VS Code.

Before this PR, we didn't log any timing information, so it was technically possible to do this logging early—we already had all the info needed.
With this PR, we want the latency and display timing logged, so we need to actually wait for the completion to be hidden first, and _then_ we can log all the data we want.

So yeah, no need to check with product because I didn't remove the logging.

Comment (olafurpg) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java:142]: Great! Just wanted to confirm 👍🏻 Feel free to merge this PR in that case

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/Autocompletion.java (hunk 7 above)
Comment (vdavid) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/Autocompletion.java:8]: Yup, I agree. I've renamed it to `AutocompleteTelemetry`.

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/Autocompletion.java (hunk 6 above)
Comment (vdavid) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/Autocompletion.java]: Added a TODO for this

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java (hunk 8 above)
Comment (vdavid) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java]: Yeah, makes sense. Autocomplete also messed up the format of this, so I've replaced it with a one-liner.

One-liner docstrings with a very broad overview of each complex method and class do make my life easier, though, so I'd love to keep adding those.

**Comment by vdavid** [comment]
> In the future, please try to keep PRs more focused on the core functional change and open separate PRs for code cleanups.

In this PR, the only refactor I did was in `GraphQlClient`, and it was closely related to the goal of this PR.
The doc additions could've gone to a different PR (maybe would've been nicer), but it seemed very low stakes so didn't want to spend time with extracting it.

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java (hunk 8 above)
Comment (olafurpg) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java]: > One-liner docstrings with a very broad overview of each complex method and class do make my life easier, though, so I'd love to keep adding those.

Agreed 👍🏻

File: client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java [hunk 9]
```diff
@@ -81,9 +110,14 @@ public void triggerAutoComplete(@NotNull Editor editor, int offset) {
             200,
             0.6,
             0.1);
+
+    // Gets AutoCompleteDocumentContext for the current offset
     TextDocument textDocument = new IntelliJTextDocument(editor, project);
     AutoCompleteDocumentContext autoCompleteDocumentContext =
         textDocument.getAutoCompleteContext(offset);
+
```
Comment (olafurpg) [inline client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java:118]: One example why unrelated changes like these can be problematic, adding this comment below introduced a merge conflict in the agent autocomplete PR because I added a TODO on the if condition

CleanShot 2023-08-11 at 11 34 15@2x


git should have been able to figure out how to merge this but it didn't

Comment (olafurpg) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java:118]: It was an easy conflict to fix, so it wasn't a big problem. It would have been worse if there were dozens of these

Comment (vdavid) [reply client/jetbrains/src/main/java/com/sourcegraph/cody/autocomplete/CodyAutoCompleteManager.java:118]: I expected conflicts on comments with that PR and wanted to tell you to feel free to discard the docs. They were more of a byproduct of my familiarizing some parts of the codebase. Sorry, forgot to mention this.
